### ISSUE TITLE
Rule proposal: no-reuse-function-inner-name

### ISSUE THREAD

**Issue by oriSomething** [eslint/eslint#6137]
There are some cases that reuse of inner function name for a param can lead to errors on some versions of some browser (some versions of safari from my experience).

this example is for this kind of problem:
```js
// example for "bad"
function error(error) {
  console.error(error);
}
```

currently i didn't find a rule that fit this case.

**Comment by mysticatea** [comment]
Thank you for this issue.

[no-shadow] rule warns such shadowing.

![image](https://cloud.githubusercontent.com/assets/1937871/15247974/824daece-1952-11e6-88f9-d2f4b89ecfb2.png)

[no-shadow]: http://eslint.org/docs/rules/no-shadow

**Comment by platinumazure** [comment]
@mysticatea I'm not sure `no-shadow` covers object properties- does it?

**Comment by ilyavolodin** [comment]
@platinumazure This issue is not about object properties (and with object properties there is no shadowing). This is about having the same name for function and arguments.
